File: parser/src/bridge.ts
/** ALP Universal Protocol Bridge (v17.0.0 — V13 The Universal Era).
 *
 * Adapts ALP objects to and from external protocol descriptions:
 *
 * - OpenAPI 3.0  — REST API specs
 * - GraphQL SDL  — schema definitions
 * - gRPC proto   — protobuf service definitions
 * - AsyncAPI     — event-driven / message API specs
 *
 * `ProtocolBridge` is the single entrypoint; `alp bridge export` and
 * `alp bridge import` are the planned CLI commands. Mirrors
 * `sdk/python/alp_sdk/bridge.py`.
 */

export const SUPPORTED_FORMATS = ['openapi', 'graphql', 'grpc', 'asyncapi'] as const
export type SupportedFormat = typeof SUPPORTED_FORMATS[number]

export class BridgeError extends Error {
  constructor(message: string) {
    super(message)
    this.name = 'BridgeError'
  }
}

export interface BridgeExportResult {
  format: SupportedFormat
  spec: Record<string, any>
  source_workflow_id: string
  warnings: string[]
}

export interface BridgeImportResult {
  format: SupportedFormat
  workflow: Record<string, any>
  source_spec: Record<string, any>
  warnings: string[]
}

export class ProtocolBridge {
  private exporters: Record<SupportedFormat, (workflow: Record<string, any>) => [Record<string, any> | string, string[]]>
  private importers: Record<SupportedFormat, (spec: any) => [Record<string, any>, string[]]>

  constructor() {
    this.exporters = {
      openapi: (wf) => this.exportOpenapi(wf),
      graphql: (wf) => this.exportGraphql(wf),
      grpc: (wf) => this.exportGrpc(wf),
      asyncapi: (wf) => this.exportAsyncapi(wf),
    }
    this.importers = {
      openapi: (spec) => this.importOpenapi(spec),
      graphql: (spec) => this.importGraphql(spec),
      grpc: (spec) => this.importGrpc(spec),
      asyncapi: (spec) => this.importAsyncapi(spec),
    }
  }

  exportWorkflow(workflow: Record<string, any>, fmt: SupportedFormat): BridgeExportResult {
    const key = fmt.toLowerCase() as SupportedFormat
    if (!(key in this.exporters)) {
      throw new BridgeError(`Unsupported export format '${fmt}'. Supported: ${SUPPORTED_FORMATS.join(', ')}`)
    }
    const [spec, warnings] = this.exporters[key](workflow)
    return {
      format: key,
      spec: spec as Record<string, any>,
      source_workflow_id: String(workflow.id ?? workflow.name ?? '_unknown'),
      warnings,
    }
  }

  importSpec(spec: Record<string, any>, fmt: SupportedFormat): BridgeImportResult {
    const key = fmt.toLowerCase() as SupportedFormat
    if (!(key in this.importers)) {
      throw new BridgeError(`Unsupported import format '${fmt}'. Supported: ${SUPPORTED_FORMATS.join(', ')}`)
    }
    const [workflow, warnings] = this.importers[key](spec)
    return {
      format: key,
      workflow,
      source_spec: spec,
      warnings,
    }
  }

  // ── OpenAPI 3.0 ─────────────────────────────────────────────────────────

  private exportOpenapi(workflow: Record<string, any>): [Record<string, any>, string[]] {
    const warnings: string[] = []
    const wfId = String(workflow.id ?? workflow.name ?? '_unknown')
    const paths: Record<string, any> = {}
    const schemas: Record<string, any> = {}
    let stepIdx = 0

    for (const step of workflow.steps ?? []) {
      const stepName = String(step.name ?? step.id ?? `step-${stepIdx++}`)
      const path = `/${stepName}`
      const requestBody = {
        content: {
          'application/json': {
            schema: { type: 'object', properties: { input: { type: 'string' } } },
          },
        },
      }
      const responses = {
        '200': {
          description: 'Success',
          content: { 'application/json': { schema: { type: 'object' } } },
        },
      }
      paths[path] = {
        post: {
          operationId: `${wfId}.${stepName}`,
          requestBody,
          responses,
        },
      }
      const schemaName = `${wfId}.${stepName}.Request`
      schemas[schemaName] = { type: 'object', properties: { input: { type: 'string' } }, required: ['input'] }
    }

    if (!Object.keys(paths).length) {
      warnings.push('Workflow has no steps; OpenAPI spec will be empty.')
    }

    const spec = {
      openapi: '3.0.0',
      info: { title: `ALP Workflow: ${wfId}`, version: '1.0.0' },
      paths,
      components: { schemas },
    }
    return [spec, warnings]
  }

  private importOpenapi(spec: Record<string, any>): [Record<string, any>, string[]] {
    const warnings: string[] = []
    const info = spec.info ?? {}
    const title = String(info.title ?? 'imported-workflow')
    const wfId = title.replace(/ /g, '-').toLowerCase()
    const steps: Array<Record<string, any>> = []

    for (const [path, methods] of Object.entries(spec.paths ?? {})) {
      const methodDetails = methods as Record<string, any>
      for (const [method, details] of Object.entries(methodDetails)) {
        if (typeof details !== 'object' || !details) continue
        const opId = String(details.operationId ?? path.replace(/^\//, ''))
        steps.push({ id: opId, name: opId, type: 'step' })
        break
      }
    }

    const workflow = {
      id: wfId,
      name: title,
      source_format: 'openapi',
      steps,
    }
    if (!steps.length) {
      warnings.push('No paths found in OpenAPI spec.')
    }
    return [workflow, warnings]
  }

  // ── GraphQL SDL ─────────────────────────────────────────────────────────

  private exportGraphql(workflow: Record<string, any>): [string, string[]] {
    const warnings: string[] = []
    const wfId = String(workflow.id ?? workflow.name ?? '_unknown').replace(/-/g, '_')
    const typeName = `${wfId}Workflow`
    const lines: string[] = [`type ${typeName} {`]
    for (const step of workflow.steps ?? []) {
      const name = String(step.name ?? step.id ?? 'step').replace(/-/g, '_')
      lines.push(`  ${name}: String`)
    }
    lines.push('}')
    lines.push('')
    lines.push(`type Query {`)
    lines.push(`  ${wfId}: ${typeName}`)
    lines.push('}')
    const sdl = lines.join('\n')
    return [sdl, warnings]
  }

  private importGraphql(spec: Record<string, any> | string): [Record<string, any>, string[]] {
    const warnings: string[] = []
    const sdl = typeof spec === 'string' ? spec : JSON.stringify(spec)
    const steps: Array<Record<string, any>> = []
    let inType = false

    for (const line of sdl.split('\n')) {
      const trimmed = line.trim()
      if (!trimmed || trimmed.startsWith('#')) continue
      if (trimmed.startsWith('type ')) {
        inType = true
        continue
      }
      if (trimmed === '}') {
        inType = false
        continue
      }
      if (inType && trimmed.includes(': ')) {
        const fieldName = trimmed.split(':')[0].trim().replace('{', '').trim()
        if (fieldName) {
          steps.push({ id: fieldName, name: fieldName, type: 'step' })
        }
      }
    }

    const workflow = {
      id: 'imported-graphql-workflow',
      name: 'Imported GraphQL Workflow',
      source_format: 'graphql',
      steps,
    }
    if (!steps.length) {
      warnings.push('No fields found in GraphQL SDL.')
    }
    return [workflow, warnings]
  }

  // ── gRPC proto ──────────────────────────────────────────────────────────

  private exportGrpc(workflow: Record<string, any>): [string, string[]] {
    const warnings: string[] = []
    const wfId = String(workflow.id ?? workflow.name ?? '_unknown').replace(/-/g, '_')
    const serviceName = `${wfId}Service`
    const lines: string[] = ['syntax = "proto3";', '', 'package alp;', `service ${serviceName} {`]

    for (const step of workflow.steps ?? []) {
      const name = String(step.name ?? step.id ?? 'step').replace(/-/g, '_')
      lines.push(`  rpc ${name}(${name}Request) returns (${name}Response);`)
    }
    lines.push('}')
    lines.push('')

    for (const step of workflow.steps ?? []) {
      const name = String(step.name ?? step.id ?? 'step').replace(/-/g, '_')
      lines.push(`message ${name}Request {`)
      lines.push('  string input = 1;')
      lines.push('}')
      lines.push(`message ${name}Response {`)
      lines.push('  string output = 1;')
      lines.push('}')
      lines.push('')
    }

    return [lines.join('\n'), warnings]
  }

  private importGrpc(spec: Record<string, any> | string): [Record<string, any>, string[]] {
    const warnings: string[] = []
    const proto = typeof spec === 'string' ? spec : JSON.stringify(spec)
    const steps: Array<Record<string, any>> = []

    for (const line of proto.split('\n')) {
      const trimmed = line.trim()
      if (trimmed.startsWith('rpc ') && trimmed.includes('(')) {
        const rpcName = trimmed.split('(')[0].replace('rpc ', '').trim()
        if (rpcName) {
          steps.push({ id: rpcName, name: rpcName, type: 'step' })
        }
      }
    }

    const workflow = {
      id: 'imported-grpc-workflow',
      name: 'Imported gRPC Workflow',
      source_format: 'grpc',
      steps,
    }
    if (!steps.length) {
      warnings.push('No RPC methods found in proto spec.')
    }
    return [workflow, warnings]
  }

  // ── AsyncAPI ────────────────────────────────────────────────────────────

  private exportAsyncapi(workflow: Record<string, any>): [Record<string, any>, string[]] {
    const warnings: string[] = []
    const wfId = String(workflow.id ?? workflow.name ?? '_unknown')
    const channels: Record<string, any> = {}

    for (const step of workflow.steps ?? []) {
      const name = String(step.name ?? step.id ?? 'step')
      const channelName = `${wfId}/${name}`
      channels[channelName] = {
        publish: {
          message: {
            name: `${name}Request`,
            payload: { type: 'object', properties: { input: { type: 'string' } } },
          },
        },
        subscribe: {
          message: {
            name: `${name}Response`,
            payload: { type: 'object', properties: { output: { type: 'string' } } },
          },
        },
      }
    }

    const spec = {
      asyncapi: '2.0.0',
      info: { title: `ALP Workflow: ${wfId}`, version: '1.0.0' },
      channels,
    }
    if (!Object.keys(channels).length) {
      warnings.push('Workflow has no steps; AsyncAPI spec will be empty.')
    }
    return [spec, warnings]
  }

  private importAsyncapi(spec: Record<string, any>): [Record<string, any>, string[]] {
    const warnings: string[] = []
    const info = spec.info ?? {}
    const title = String(info.title ?? 'imported-asyncapi-workflow')
    const wfId = title.replace(/ /g, '-').toLowerCase()
    const steps: Array<Record<string, any>> = []

    for (const channel of Object.values(spec.channels ?? {})) {
      if (typeof channel === 'object' && channel) {
        const pub = (channel as Record<string, any>).publish ?? {}
        const msg = pub.message ?? {}
        const name = String(msg.name ?? 'step')
        steps.push({ id: name, name: name, type: 'step' })
      }
    }

    const workflow = {
      id: wfId,
      name: title,
      source_format: 'asyncapi',
      steps,
    }
    if (!steps.length) {
      warnings.push('No channels found in AsyncAPI spec.')
    }
    return [workflow, warnings]
  }
}

File: sdk/python/alp_sdk/bridge.py
"""ALP Universal Protocol Bridge (v17.0.0 — V13 The Universal Era).

Adapts ALP objects to and from external protocol descriptions:

* OpenAPI 3.0  — REST API specs
* GraphQL SDL  — schema definitions
* gRPC proto   — protobuf service definitions
* AsyncAPI     — event-driven / message API specs

``ProtocolBridge`` is the single entrypoint; ``alp bridge export`` and
``alp bridge import`` are the planned CLI commands. Tests live in
``sdk/python/tests/test_bridge.py``.
"""
from __future__ import annotations


import json
import textwrap
from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional, Union


SUPPORTED_FORMATS = ["openapi", "graphql", "grpc", "asyncapi"]


class BridgeError(Exception):
    """Raised when a bridge conversion cannot be performed."""


@dataclass
class BridgeExportResult:
    format: str
    spec: Dict[str, Any]
    source_workflow_id: str
    warnings: List[str] = field(default_factory=list)

    def to_dict(self) -> Dict[str, Any]:
        return {
            "format": self.format,
            "spec": self.spec,
            "source_workflow_id": self.source_workflow_id,
            "warnings": self.warnings,
        }


@dataclass
class BridgeImportResult:
    format: str
    workflow: Dict[str, Any]
    source_spec: Dict[str, Any]
    warnings: List[str] = field(default_factory=list)

    def to_dict(self) -> Dict[str, Any]:
        return {
            "format": self.format,
            "workflow": self.workflow,
            "source_spec": self.source_spec,
            "warnings": self.warnings,
        }


class ProtocolBridge:
    """Bidirectional adapter between ALP objects and external API specs.

    Usage::

        bridge = ProtocolBridge()
        result = bridge.export_workflow(workflow, fmt="openapi")
        spec = result.to_dict()["spec"]

        result = bridge.import_spec(spec, fmt="openapi")
        workflow = result.to_dict()["workflow"]
    """

    def __init__(self):
        self._exporters = {
            "openapi": self._export_openapi,
            "graphql": self._export_graphql,
            "grpc": self._export_grpc,
            "asyncapi": self._export_asyncapi,
        }
        self._importers = {
            "openapi": self._import_openapi,
            "graphql": self._import_graphql,
            "grpc": self._import_grpc,
            "asyncapi": self._import_asyncapi,
        }

    def export_workflow(self, workflow: Dict[str, Any], fmt: str) -> BridgeExportResult:
        fmt = fmt.lower()
        if fmt not in self._exporters:
            raise BridgeError(f"Unsupported export format '{fmt}'. Supported: {SUPPORTED_FORMATS}")
        exporter = self._exporters[fmt]
        spec, warnings = exporter(workflow)
        return BridgeExportResult(format=fmt, spec=spec, source_workflow_id=workflow.get("id", "_unknown"), warnings=warnings)

    def import_spec(self, spec: Dict[str, Any], fmt: str) -> BridgeImportResult:
        fmt = fmt.lower()
        if fmt not in self._importers:
            raise BridgeError(f"Unsupported import format '{fmt}'. Supported: {SUPPORTED_FORMATS}")
        importer = self._importers[fmt]
        workflow, warnings = importer(spec)
        return BridgeImportResult(format=fmt, workflow=workflow, source_spec=spec, warnings=warnings)

    # ── OpenAPI 3.0 ─────────────────────────────────────────────────────────

    def _export_openapi(self, workflow: Dict[str, Any]) -> tuple[Dict[str, Any], List[str]]:
        warnings: List[str] = []
        wf_id = workflow.get("id", workflow.get("name", "_unknown"))
        paths: Dict[str, Any] = {}
        schemas: Dict[str, Any] = {}
        step_idx = 0

        for step in workflow.get("steps", []):
            step_name = str(step.get("name", step.get("id", f"step-{step_idx}")))
            path = f"/{step_name}"
            step_idx += 1
            request_body = {"content": {"application/json": {"schema": {"type": "object", "properties": {"input": {"type": "string"}}}}}}
            responses = {"200": {"description": "Success", "content": {"application/json": {"schema": {"type": "object"}}}}}
            paths[path] = {"post": {"operationId": f"{wf_id}.{step_name}", "requestBody": request_body, "responses": responses}}
            schema_name = f"{wf_id}.{step_name}.Request"
            schemas[schema_name] = {"type": "object", "properties": {"input": {"type": "string"}}, "required": ["input"]}

        if not paths:
            warnings.append("Workflow has no steps; OpenAPI spec will be empty.")

        spec = {
            "openapi": "3.0.0",
            "info": {"title": f"ALP Workflow: {wf_id}", "version": "1.0.0"},
            "paths": paths,
            "components": {"schemas": schemas},
        }
        return spec, warnings

    def _import_openapi(self, spec: Dict[str, Any]) -> tuple[Dict[str, Any], List[str]]:
        warnings: List[str] = []
        info = spec.get("info", {})
        title = info.get("title", "imported-workflow")
        wf_id = title.replace(" ", "-").lower()
        steps: List[Dict[str, Any]] = []
        for path, methods in spec.get("paths", {}).items():
            for method, details in methods.items():
                if not isinstance(details, dict):
                    continue
                op_id = details.get("operationId", path.strip("/"))
                steps.append({"id": op_id, "name": op_id, "type": "step"})
                break
        workflow = {
            "id": wf_id,
            "name": title,
            "source_format": "openapi",
            "steps": steps,
        }
        if not steps:
            warnings.append("No paths found in OpenAPI spec.")
        return workflow, warnings

    # ── GraphQL SDL ─────────────────────────────────────────────────────────

    def _export_graphql(self, workflow: Dict[str, Any]) -> tuple[str, List[str]]:
        warnings: List[str] = []
        wf_id = workflow.get("id", workflow.get("name", "_unknown")).replace("-", "_")
        type_name = f"{wf_id}Workflow"
        lines = [f"type {type_name} {{"]
        for step in workflow.get("steps", []):
            name = str(step.get("name", step.get("id", "step"))).replace("-", "_")
            lines.append(f"  {name}: String")
        lines.append("}")
        lines.append("")
        lines.append(f"type Query {{")
        lines.append(f"  {wf_id}: {type_name}")
        lines.append("}")
        sdl = "\n".join(lines)
        return sdl, warnings

    def _import_graphql(self, spec: Union[str, Dict[str, Any]]) -> tuple[Dict[str, Any], List[str]]:
        warnings: List[str] = []
        if isinstance(spec, dict):
            sdl = json.dumps(spec)
        else:
            sdl = str(spec)
        steps: List[Dict[str, Any]] = []
        for line in sdl.splitlines():
            line = line.strip()
            if not line or line.startswith("type ") or line.startswith("}"):
                continue
            if ": " in line and not line.startswith("#"):
                field_name = line.split(":")[0].strip().strip("{")
                if field_name:
                    steps.append({"id": field_name, "name": field_name, "type": "step"})
        workflow = {
            "id": "imported-graphql-workflow",
            "name": "Imported GraphQL Workflow",
            "source_format": "graphql",
            "steps": steps,
        }
        if not steps:
            warnings.append("No fields found in GraphQL SDL.")
        return workflow, warnings

    # ── gRPC proto ──────────────────────────────────────────────────────────

    def _export_grpc(self, workflow: Dict[str, Any]) -> tuple[str, List[str]]:
        warnings: List[str] = []
        wf_id = workflow.get("id", workflow.get("name", "_unknown")).replace("-", "_")
        service_name = f"{wf_id}Service"
        lines = [f'syntax = "proto3";', "", f"package alp;", "", f"service {service_name} {{"]
        for step in workflow.get("steps", []):
            name = str(step.get("name", step.get("id", "step"))).replace("-", "_")
            lines.append(f"  rpc {name}({name}Request) returns ({name}Response);")
        lines.append("}")
        lines.append("")
        for step in workflow.get("steps", []):
            name = str(step.get("name", step.get("id", "step"))).replace("-", "_")
            lines.append(f"message {name}Request {{")
            lines.append(f"  string input = 1;")
            lines.append(f"}}")
            lines.append(f"message {name}Response {{")
            lines.append(f"  string output = 1;")
            lines.append(f"}}")
            lines.append("")
        proto = "\n".join(lines)
        return proto, warnings

    def _import_grpc(self, spec: Union[str, Dict[str, Any]]) -> tuple[Dict[str, Any], List[str]]:
        warnings: List[str] = []
        if isinstance(spec, dict):
            proto = json.dumps(spec)
        else:
            proto = str(spec)
        steps: List[Dict[str, Any]] = []
        for line in proto.splitlines():
            line = line.strip()
            if line.startswith("rpc ") and "(" in line:
                rpc_name = line.split("(")[0].replace("rpc ", "").strip()
                if rpc_name:
                    steps.append({"id": rpc_name, "name": rpc_name, "type": "step"})
        workflow = {
            "id": "imported-grpc-workflow",
            "name": "Imported gRPC Workflow",
            "source_format": "grpc",
            "steps": steps,
        }
        if not steps:
            warnings.append("No RPC methods found in proto spec.")
        return workflow, warnings

    # ── AsyncAPI ────────────────────────────────────────────────────────────

    def _export_asyncapi(self, workflow: Dict[str, Any]) -> tuple[Dict[str, Any], List[str]]:
        warnings: List[str] = []
        wf_id = workflow.get("id", workflow.get("name", "_unknown"))
        channels: Dict[str, Any] = {}
        for step in workflow.get("steps", []):
            name = str(step.get("name", step.get("id", "step")))
            channel_name = f"{wf_id}/{name}"
            channels[channel_name] = {
                "publish": {"message": {"name": f"{name}Request", "payload": {"type": "object", "properties": {"input": {"type": "string"}}}}},
                "subscribe": {"message": {"name": f"{name}Response", "payload": {"type": "object", "properties": {"output": {"type": "string"}}}}},
            }
        spec = {
            "asyncapi": "2.0.0",
            "info": {"title": f"ALP Workflow: {wf_id}", "version": "1.0.0"},
            "channels": channels,
        }
        if not channels:
            warnings.append("Workflow has no steps; AsyncAPI spec will be empty.")
        return spec, warnings

    def _import_asyncapi(self, spec: Dict[str, Any]) -> tuple[Dict[str, Any], List[str]]:
        warnings: List[str] = []
        info = spec.get("info", {})
        title = info.get("title", "imported-asyncapi-workflow")
        wf_id = title.replace(" ", "-").lower()
        steps: List[Dict[str, Any]] = []
        for channel in spec.get("channels", {}).values():
            if isinstance(channel, dict):
                pub = channel.get("publish", {})
                msg = pub.get("message", {})
                name = msg.get("name", "step")
                steps.append({"id": name, "name": name, "type": "step"})
        workflow = {
            "id": wf_id,
            "name": title,
            "source_format": "asyncapi",
            "steps": steps,
        }
        if not steps:
            warnings.append("No channels found in AsyncAPI spec.")
        return workflow, warnings
